fix(jqLite) - toggleClass & removeClass reflect documentation

JordanMajd opened the pull request:
Updates removeClass to accept function  in spec with 1.9 (current) jQuery docs.
Updates removeClass to accept no parameters in spec with 1.9 (current) jQuery docs.
Updates toggleClass to accept function in spec with 1.9 (current) jQuery docs.
Updates toggleClass to accept no parameters in spec with 1.9 (current) jQuery docs.
Provides tests for all changes.

fixes angular/angular.js#12793
fixes angular/angular.js#12851

JordanMajd committed:
fix(jqLite) - toggleClass & removeClass reflect documentation

toggleClass & removeClass changes make functions reflect documentation and jQuery 1.4 compatible.
toggleClass now can accept a function as a parameter, a function and array of states as a parameter
as well as empty parameters. removeClass now can accept a function as a parameter as well as
empty parameters.

closes angular/angular.js#12793
closes angular/angular.js#12851

Narretz commented on src/jqLite.js:
```diff
@@ -381,15 +381,26 @@ function jqLiteHasClass(element, selector) {
+function jqLiteRemoveClass(element, cssClassesOrFunc) {
+  if (element.setAttribute) {
+    if (cssClassesOrFunc) {
+      if (typeof cssClassesOrFunc === 'function') {
```
can use `isFunction`

Narretz commented on src/jqLite.js:
```diff
@@ -381,15 +381,26 @@ function jqLiteHasClass(element, selector) {
+  if (element.setAttribute) {
+    if (cssClassesOrFunc) {
+      if (typeof cssClassesOrFunc === 'function') {
+        forEach(element.className.split(/\s+/), function(className, index) {
```
I don't think this is doing the right thing. jquer docs say:
> A function returning one or more space-separated class names to be added to the existing class name(s). Receives the index position of the element in the set and the existing class name(s) as arguments. Within the function, this refers to the current element in the set.

JordanMajd replied:
@Narretz 
I will look into this, thank you.

Narretz commented on src/jqLite.js:
```diff
@@ -935,14 +946,35 @@ forEach({
-      });
+
+      // the class name to be toggled can be determined by passing in a function.
+      if (typeof selector === 'function') {
```
-> `isFunction`

Narretz commented on src/jqLite.js:
```diff
@@ -935,14 +946,35 @@ forEach({
+      if (typeof selector === 'function') {
+
+        forEach(element.className.split(/\s+/), function(className, index) {
+          if (Array.isArray(condition) && condition.length >= index) {
```
-> `isArray`

JordanMajd committed:
fix(jqLite) removeClass now compatible with jQuery 1.4

remove class is matches documentation and is jQuery 1.4 compatible. Changes all function and array
checks to use isArray and isFunction.

JordanMajd committed:
fix(jqLite): removeClass & toggleClass matches jQuery 1.9 spec

Updates removeClass to accept function  in spec with 1.9 (current) jQuery docs.
Updates removeClass to accept no parameters in spec with 1.9 (current) jQuery docs.
Updates toggleClass to accept function in spec with 1.9 (current) jQuery docs.
Updates toggleClass to accept no parameters in spec with 1.9 (current) jQuery docs.
Provides tests for all changes.

JordanMajd commented:
@Narretz , thanks for the reviews. I've fixed the items you had mentioned as well as a few other things I had noticed that could use improvement and am ready for another review.

gkalpak commented:
There is still the inconsistency of the passed index. (It's always 0.)

JordanMajd commented:
Thanks for the input. However, I think passing anything but 0 would be inconsistent. Please correct me if I'm mistaken, but it is my understanding that  `jqLite` and `jQuery` are intrinsically going to be different. This is because `jQuery` can work on multiple elements, while `jqLite` will only operate on a single element. Since `jqLite` is only operating on one element, the element will always be index 0.

gkalpak commented:
Off the top of my head, I think `jqLite` operates on multiple elements as well.

JordanMajd commented:
@gkalpak thanks, 

I just did some digging, you are correct! I will fix this.

Narretz commented:
I'm going to close this PR because we cannot merge it in its current state, as we are either waiting for more information, tests are missing, or required code changes haven't been made. 
Feel free to reopen this PR or open a new one if you have remedied the issues pointed out.
